### ISSUE TITLE
AndroidX migration

### ISSUE THREAD

**Issue by baksansk** [Adobe-Marketing-Cloud/acp-sdks#602]
Please migrate the Adobe Android SDKs to use AndroidX libraries instead of the old Android support libraries. See migration guide here: https://developer.android.com/jetpack/androidx/migrate

Although developers can consume your libraries by using the AndroidX Jetifier (transforming your AARs to replace android.support references with their equivalent androidx classes), it is considered slow and affects incremental build time.
Many Android dev teams that are build conscious would want to disable AndroidX Jetifier to improve build time and would blocked by Adobe's SDKs if they use them. Although some workarounds exist, it would be best if you can migrate to AndroidX.

By using this Gradle plugin, I'm able to find the following references in your AARs to the old Android support libraries.
https://medium.com/dipien/say-bye-bye-to-android-jetifier-a7e0d388f5d6
https://github.com/dipien/bye-bye-jetifier

`
com.adobe.marketing.mobile:assurance:1.0.0
 Issues found:
 * com/adobe/marketing/mobile/AssuranceClientInfo.class -> android/support/v4/app/ActivityCompat
 * com/adobe/marketing/mobile/AssuranceFloatingButtonView.class -> android/support/v4/content/ContextCompat
`
Your pom file also indicates the dependency on the old support libs: https://search.maven.org/artifact/com.adobe.marketing.mobile/assurance/1.0.3/aar

com.adobe.marketing.mobile:core:1.5.7
 Issues found:
 * com/adobe/marketing/mobile/LocalNotificationHandler.class -> android.support.v4.app.NotificationCompat$Builder
 * com/adobe/marketing/mobile/LocalNotificationHandler.class -> android.support.v4.app.NotificationCompat$Style
 * com/adobe/marketing/mobile/LocalNotificationHandler.class -> android.support.v4.app.NotificationCompat$BigTextStyle

**Comment by ahmyr** [comment]
@baksansk, our current minimum for Android support is API 14. AndroidX requires API 15. We will look into bumping the minimum version and supporting AndroidX.

Amber

**Comment by svenjacobs** [comment]
Please think about prioritizing this. We had to enable [jetifier](https://developer.android.com/studio/command-line/jetifier) – which slows down build significantly – only for Adobe Assurance. API 15 as minimum supported SDK is recommended by Google since 2014! Thank you 😃

**Comment by svenjacobs** [comment]
@ahmyr Do you have any news for us, please?

**Comment by JohannesPtaszyk** [comment]
We also face issues with this missing migration.

**Comment by sky-ricardocarrapico** [comment]
Same issue on our side. It's the only SDK that forces us to enable Jetifier.

**Comment by G00fY2** [comment]
> @svenjacobs, we have added it to our backlog, but no ETA yet. We'll update here once we have it available.

Is there any technical limitation that prevents you from addressing this or does it just have low priority? Because the actual migration (link to the guide is posted above) on your side is pretty straight forward and should at most take a few hours.

**Comment by joakimcarlgren** [comment]
Is there any update on this? Any bump in the road, when having to add an analytics tool suggested by marketing, will lead to pushback from developers. This feels like a significant bump in the road, so would be good to understand if and when this will be fixed.

**Comment by JohannesPtaszyk** [comment]
> @svenjacobs, we have added it to our backlog, but no ETA yet. We'll update here once we have it available.

It has been another 4 months now and the actual migration should only take a couple of hours. 
AndroidX has been out for a really long time and it is getting somewhat frustrating that this issue is not getting tackled with a steadily growing number of requests. 

Another update from your side would be really appreciated.

**Comment by aya2453** [comment]
Same here

**Comment by spoorthipujariadobe** [comment]
We are currently working on this. We do not have an ETA but will work towards releasing it soon.

Thanks,
Spoorthi

**Comment by emartynov** [comment]
AGP 7.2 warns us about the enabled jetifier now - it is a warning in the build now. And this library the only one that still requires it.

**Comment by svenjacobs** [comment]
I usually refrain from posting "Any news" comments but since the Adobe SDK is probably the last library for most of us that requires Jetifier to be enabled, since Android Studio Chipmunk now warns about this, since this ticket is already 9 months old and since the migration should only take a few hours at most and therefore this whole situation is very frustrating, I think I speak for all of us when I ask, when can we expect a release?

**Comment by emartynov** [comment]
Agree, I would do migration myself if the code is public

**Comment by emartynov** [comment]
It is time to ask again - any updates? Any ETA? Any hope?

**Comment by ahmyr** [comment]
@emartynov, we have started work on this, but no ETA on the moment.

**Comment by emartynov** [comment]
Thankfully we have company maven now. In 15 minutes I made a jetified library version and uploaded it to our maven. So finally, I turned off jetifier completely.

**Comment by BlackHornet** [comment]
> Thankfully we have company maven now. In 15 minutes I made a jetified library version and uploaded it to our maven. So finally, I turned off jetifier completely.

Not a solution for everybody 🤡

**Comment by G00fY2** [comment]
Happy 1st anniversary of this ticket. 🥳

I don't like to "spam" GitHub issues this way but are there any updates about the AndroidX migration? Again 2 month have passed since you stated that you are working on it. I was hoping your recently released [minor update of the affected Adobe target](https://aep-sdks.gitbook.io/docs/release-notes#august-2-2022) library would contain any changes regarding the migration, but unfortunately it still references old support libraries.

**Comment by praschetan** [comment]
We're currently on a major refactoring effort of our Android libraries and among the changes coming are: improved stability, equivalent features to iOS Swift libraries (configuration, registration, etc.) and support for Android X. Apologies for the delays - currently, we're looking at a Q1, 2023 release timeframe.

**Comment by svenjacobs** [comment]
@praschetan Thank you for your detailed explanation. I understand that a major refactoring is a huge effort. However in this issue it was mentioned several times that the AndroidX refactoring should only take a few hours since it's basically a swap of libraries and renaming of packages. Q1 2023 is still over half a year away. So if you could spare a few hours and migrate the existing library, you would make a lot of us happy. Thank you very much.

**Comment by JohannesPtaszyk** [comment]
I'd also like to thank you for the update, but I also agree with @svenjacobs. This is really not a huge effort and it would make so many developers more happy.

**Comment by JohannesPtaszyk** [comment]
Agreed @svenjacobs :)

**Comment by svenjacobs** [comment]
> currently, we're looking at a Q1, 2023 release timeframe.

@praschetan @spoorthipujariadobe @ahmyr 

It is Q1 2023 now. Do you please have some positive news for us? 🙏🏼 We would like to finally get rid of jetifier. This issue is now almost 1.5 years old. The migration to AndroidX should only take a few hours for you. Imagine how many developer hours you could save by this change, because build times increase significantly when jetifier is enabled.

**Comment by JohannesPtaszyk** [comment]
@svenjacobs We finally decided to jetify the latest release manually once and include the resulting lib in the libs folder of our project. 
Not the nicest thing to do, but we could finally turn off jetifier.

**Comment by svenjacobs** [comment]
> We finally decided to jetify the latest release manually once and include the resulting lib in the libs folder of our project.

I now did this, too, but it was a bit more complicated. Since Adobe Assurance was a dependency of an Android library, I first got the error

```
Direct local .aar file dependencies are not supported when building an AAR
```

So I had to move all the Adobe code into the Android application (`main` module). All this because Adobe is unable to migrate their library to AndroidX 🙄 I will certainly remember how Adobe treats their customers / developers and share my bad experience 😡 

If someone else wants to go this route, too, the dependency can be downloaded [here](https://github.com/Adobe-Marketing-Cloud/acp-sdks/blob/master/android/assurance-1.0.4.aar) and Jetifier is found [here](https://developer.android.com/studio/command-line/jetifier).

**Comment by svenjacobs** [comment]
One more obstacle I faced after adding the manually jetified library to my `main` module was that the assurance artifact declares an `AppTheme` style, which is a common name for app themes. They don't prefix their resources, which is recommended for Android library developers 🙄 So their theme was overwriting my app's theme and the whole design was broken. I had to rename my theme in order to fix this.

**Comment by JohannesPtaszyk** [comment]
Had the same with app_name once :D

**Comment by G00fY2** [comment]
Looks like [assurance](https://github.com/adobe/aepsdk-assurance-android/blob/main/code/gradle.properties#L30) and [target](https://github.com/adobe/aepsdk-target-android/blob/dev/code/gradle.properties#L26) now use AndroidX in version 2.0.

Will crosspost this question here: https://github.com/adobe/aepsdk-core-android/issues/394

Without a proper changelog or migration guide we are blocked from updating.

**Comment by ahmyr** [comment]
AndroidX support has been added :) . Please see the info here:
Release notes: https://developer.adobe.com/client-sdks/documentation/release-notes/#february-23-2023
Migration Guide: https://developer.adobe.com/client-sdks/previous-versions/documentation/migrate-to-android/

**Comment by svenjacobs** [comment]
Although I don't agree on how this issue was handled and wish that Adobe improves communication with their developer community in the future, I'm glad that 2.0 with AndroidX support was finally released. I try to migrate to the new SDK as soon as possible.
